─────────────────────────────────────────────
* REAL DATA — MacBook Pro 14-inch official USD launch prices
* Source: Apple.com / MacRumors / AppleInsider

* Chip  RAM   Price ($)

─────────────────────────────────────────────
 * M1     16    1999
 * M1     32    2499   (M1 Pro)
 * M2     16    1999
 * M2     24    2199   (M2 Pro)
 * M2     32    2399   (M2 Pro)
 * M3     16    1599
 * M3     24    1999   (M3 Pro)
 * M3     32    2399   (M3 Pro)
 * M4     16    1599
 * M4     24    1999   (M4 Pro)
 * M4     32    2399   (M4 Pro)
 * M4     48    2799   (M4 Pro)
 * M5     16    1699
 * M5     24    1899
 * M5     32    2099
 * M5     48    3099   (M5 Pro)
 
─────────────────────────────────────────────
* Feature 1: Chip generation (M1=1, M2=2, M3=3, M4=4, M5=5)
* Feature 2: RAM in GB

In [1]:
import numpy as np 
chip_gen = np.array([1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5], dtype=float)
ram      = np.array([16, 32, 16, 24, 32, 16, 24, 32, 16, 24, 32, 48, 16, 24, 32, 48], dtype=float)
prices   = np.array([1999, 2499, 1999, 2199, 2399, 1599, 1999, 2399,
                     1599, 1999, 2399, 2799, 1699, 1899, 2099, 3099], dtype=float)

In [5]:
# Design Matrix X
ones=np.ones((len(prices),1))
X=np.hstack([ones,chip_gen.reshape(-1,1),ram.reshape(-1,1)])
print("Designed Matrix X: \n",X)
print("Shape: ",X.shape)

Designed Matrix X: 
 [[ 1.  1. 16.]
 [ 1.  1. 32.]
 [ 1.  2. 16.]
 [ 1.  2. 24.]
 [ 1.  2. 32.]
 [ 1.  3. 16.]
 [ 1.  3. 24.]
 [ 1.  3. 32.]
 [ 1.  4. 16.]
 [ 1.  4. 24.]
 [ 1.  4. 32.]
 [ 1.  4. 48.]
 [ 1.  5. 16.]
 [ 1.  5. 24.]
 [ 1.  5. 32.]
 [ 1.  5. 48.]]
Shape:  (16, 3)


In [6]:
# Checking invertibility
Xt=np.transpose(X)
XtX=Xt @ X
det_xtx=np.linalg.det(XtX)
print("Result: ",det_xtx)

Result:  724223.9999999995


Doesn't equals to zero. So predication can be done.

In [ ]:
# Normal Equation
w=np.linalg.inv(XtX) @ Xt @ prices
print("Result: ",w)

Result:  [1358.87981619  -78.0311064    39.53141569]


* Intercept (bias term) = 1358.88  
→ This is the baseline price (in dollars) when both chip generation and RAM are set to zero. Of course, such a MacBook doesn’t exist, but mathematically it’s the starting point of the regression line.

* Coefficient for chip generation = -78.03  
→ For each additional chip generation (M1 → M2 → M3, etc.), the model predicts the price decreases by about $78. This reflects Apple’s repositioning of base models (e.g., M3/M4 dropped to $1599), so the trend is slightly negative.

* Coefficient for RAM = 39.53  
→ For each additional GB of RAM, the price increases by about $39.50. This is the dominant driver of price in the model, showing how memory upgrades scale cost.

In [10]:
# Predictions on Training Data.
y_pred=X @ w
print("Predicted prices for each macbook configuration: \n",y_pred)

Predicted prices for each macbook configuration: 
 [1913.3513609  2545.85401202 1835.32025451 2151.57158006 2467.82290562
 1757.28914811 2073.54047367 2389.79179922 1679.25804171 1995.50936727
 2311.76069282 2944.26334394 1601.22693531 1917.47826087 2233.72958643
 2866.23223754]


In [12]:
# MSE & RMSE
mse=np.mean((prices-y_pred)**2)
rmse=np.sqrt(mse)
print("Mean Square Error: ",mse)
print("Root Mean Square Error: ",rmse)

Mean Square Error:  11962.762902085553
Root Mean Square Error:  109.37441612226121


* Mean Square Error (MSE) = 11,962.76  
→ This is the average of the squared differences between actual and predicted prices. It measures overall model fit, with larger errors penalized more heavily. A lower MSE indicates better accuracy.

* Root Mean Square Error (RMSE) = 109.37  
→ This is the square root of MSE, expressed in the same units as the target variable (dollars). It represents the typical prediction error size. In this case, the model’s average miss is about $109 per MacBook configuration.